### PR TITLE
Robot diagnostics + Robot software robocup cleanup

### LINKED ISSUE #1201: Visual Indication for when e-stop is on (RobotDiagnostics)
### Description of the task

After several occurrences of thinking robots were broken when the e-stop was just on without us realizing, Jade made the excellent suggestion that we give a Visual cue in our GUI tools that indicate the e-stop is on. 

Options are:
* Make everything red (but still readable)
* Something else? Open to ideas

### Acceptance criteria

- [ ] Add a visual indication to the Robot Diagnostics tool and AI Visualizer that indicate when the e-stop is on

### LINKED ISSUE #2563: Thunderscope: Robot Diagnostics Integration

### LINKED ISSUE #2594: Flashing ESP32 from the Jetson Nano
We need to flash the esp32 from the jetson over the uart connection we use to communicate with power loop.

### LINKED ISSUE #2644: Integrate RobotStatus msgs with Thunderscope (per-wheel velocities, target velocity, temps, etc..)

### LINKED ISSUE #2729: Return NetworkStatus when NetworkService is polled
Should return NetworkStatus when we poll NetworkService

### LINKED ISSUE #2730: Return DriveUnitStatus when MotorService is polled
Poll MotorStatus in Thunderloop

### PR DESCRIPTION
**Pull request by LiCody**
### Description

- Missing
  - HRVO changes from robocup
  - Proto Serialization changes
- Includes
  - Robot Diagnostics
  - Other Robot Software
  - Motor Recovery code
  - powerloop + power service changes + per robot kick/chip speed

### Testing Done

Tested with robot

### Resolved Issues

resolves https://github.com/UBC-Thunderbots/Software/issues/2594
resolves https://github.com/UBC-Thunderbots/Software/issues/1201
resolves https://github.com/UBC-Thunderbots/Software/issues/2644
resolves https://github.com/UBC-Thunderbots/Software/issues/2563
resolves #2729 
resolves #2730 

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue

### CODE REVIEW

File: src/shared/2021_robot_constants.cpp [hunk 1]
```diff
@@ -10,20 +10,21 @@ RobotConstants_t create2021RobotConstants(void)
...
         // TODO (#2112): update this
         .jerk_limit_kg_m_per_s_3 = 40.0f,
-        .front_wheel_angle_deg   = 57.95f,
-        .back_wheel_angle_deg    = 136.04f,
+        .front_wheel_angle_deg =
+            34.06f,  // The measured angle is 32 degrees, but it only works at 34
+        .back_wheel_angle_deg = 46.04f,
         // TODO (#2112): update this
         .front_of_robot_width_meters = 0.11f,
         // TODO (#2112): update this
```
Comment (itsarune) [inline src/shared/2021_robot_constants.cpp:18]: either leave #2112 open or remove the TODOs

File: src/software/ai/hl/stp/play/penalty_kick_enemy_play_test.cpp [hunk 2]
```diff
@@ -25,6 +25,7 @@ class PenaltyKickEnemyPlayTest
     Field field                      = Field::createField(field_type);
 };
 
+/** TODO (#2714): Re-enable tests
 TEST_P(PenaltyKickEnemyPlayTest, test_penalty_kick_enemy_play_setup)
```
Comment (itsarune) [inline src/software/ai/hl/stp/play/penalty_kick_enemy_play_test.cpp]: the syntax we use to disable a test:
`TEST_P(PenaltyKickEnemyPlayTest, DISABLED_test_penalty_kick_enemy_play_setup`

File: src/software/ai/hl/stp/tactic/shadow_enemy/shadow_enemy_tactic_test.cpp [hunk 3]
```diff
@@ -119,6 +119,7 @@ TEST_F(ShadowEnemyTacticTest, test_block_pass_if_enemy_does_not_have_ball)
             Duration::fromSeconds(5));
 }
 
+/** TODO (#2714): Re-enable tests
 TEST_F(ShadowEnemyTacticTest, test_block_net_then_steal_and_chip)
```
Comment (itsarune) [inline src/software/ai/hl/stp/tactic/shadow_enemy/shadow_enemy_tactic_test.cpp]: update syntax for disabling a test

File: src/software/jetson_nano/services/network.cpp [hunk 4]
```diff
@@ -34,10 +37,28 @@ void NetworkService::primitiveSetCallback(TbotsProto::PrimitiveSet input)
...
+    }
+
+    float packet_loss = static_cast<float>(total_primitives_lost) /
+                        static_cast<float>(input.sequence_number());
+
+    if (packet_loss > PACKET_LOSS_WARNING_THRESHOLD)
+    {
+        LOG(WARNING) << "Primitive packet loss more than "
+                     << PACKET_LOSS_WARNING_THRESHOLD << "% ";
+    }
```
Comment (itsarune) [inline src/software/jetson_nano/services/network.cpp:51]: this is an averaged count, we should make a count that checks how much packet loss we're seeing in the last 100 or so packets

you can tag an issue for this or implement it yourself

**Review by itsarune — COMMENTED** [review]
thanks for taking this up Cody, I have a few comments + there's merge conflict with master

File: src/shared/2021_robot_constants.cpp [hunk 5]
```diff
@@ -10,20 +10,21 @@ RobotConstants_t create2021RobotConstants(void)
...
         .robot_max_ang_speed_rad_per_s          = 56.76f,
-        .robot_max_acceleration_m_per_s_2       = 3.28f,
-        .robot_max_ang_acceleration_rad_per_s_2 = 38.57f,
+        .robot_max_acceleration_m_per_s_2       = 4.0f,
+        .robot_max_ang_acceleration_rad_per_s_2 = 20.0f,
         // TODO (#2112): update this
         .indefinite_dribbler_speed_rpm = 10000.0f,
         // TODO (#2112): update this
-        .max_force_dribbler_speed_rpm       = 16000.0f,
+        .max_force_dribbler_speed_rpm       = -12000.0f,
```
Comment (nimazareian) [inline src/shared/2021_robot_constants.cpp:28]: Max being a negative number seems counter-intuitive. Should we possibly negate the value wherever it's used in the implementation?

File: src/software/jetson_nano/display/robot_ui.py [hunk 6]
```diff
@@ -35,21 +31,11 @@ class ScreenActions:
...
-    "wheels enable",
-    "fl wheel speed",
-    "fr wheel speed",
-    "bl wheel speed",
-    "br wheel speed",
+    "/robot_id",
+    "/channel_id",
+    "/battery_voltage",
+    "/cap_voltage",
+    "/current_draw",
```
Comment (nimazareian) [inline src/software/jetson_nano/display/robot_ui.py:39]: nit: We should use the constants defined in `software/constants.h` instead of copying the values.

File: src/software/jetson_nano/display/screens/home_screen.py [hunk 7]
```diff
@@ -21,15 +21,15 @@ def __init__(self, lcd_display, redis_dict, screen_actions):
         """
         actions = [
             {
-                "redis key": "robot id",
-                "value": redis_dict["robot id"],
+                "redis key": "/robot_id",
```
Comment (nimazareian) [inline src/software/jetson_nano/display/screens/home_screen.py:25]: The constants should also be used in this file as well.

File: src/software/jetson_nano/services/motor.h [hunk 8]
```diff
@@ -189,5 +212,8 @@ class MotorService
     EuclideanToWheel euclidean_to_four_wheel;
     std::unordered_map<int, bool> encoder_calibrated_;
 
-    int heartbeat_state = 0;
+    // Previous wheel velocities
+    WheelSpace_t prev_wheel_velocities;
+
+    int count = 0;
```
Comment (nimazareian) [inline src/software/jetson_nano/services/motor.h]: I don't think `count` or `previous_dribbler_rpm` are being used anymore.

File: src/software/jetson_nano/services/network.cpp [hunk 9]
```diff
@@ -34,10 +37,28 @@ void NetworkService::primitiveSetCallback(TbotsProto::PrimitiveSet input)
 {
     std::scoped_lock<std::mutex> lock(primitive_set_mutex);
     primitive_set_msg = input;
+
+    if (last_primitive_sequence_number != input.sequence_number() - 1)
+    {
+        total_primitives_lost++;
```
Comment (nimazareian) [inline src/software/jetson_nano/services/network.cpp]: Should this not be the following so we count all the primitives lost (Assuming that there could be more than between two callbacks): 
```cpp
total_primitives_lost += input.sequence_number() - last_primitive_sequence_number - 1
```
If so, we can remove the if statement as well.

File: src/software/jetson_nano/services/network.h [hunk 10]
```diff
@@ -53,4 +53,10 @@ class NetworkService
     // Functions to callback primitiveSet and world and stores them in a variable
     void primitiveSetCallback(TbotsProto::PrimitiveSet input);
     void worldCallback(TbotsProto::World input);
+
+    double last_primitive_time = 0.0;
+    double last_world_time     = 0.0;
```
Comment (nimazareian) [inline src/software/jetson_nano/services/network.h:58]: Neither of these two values are really being used. We probably should remove them and implement something like what we have for lost primitive sets. Could be a good level 3 task!

File: src/software/jetson_nano/thunderloop.cpp [hunk 11]
```diff
@@ -194,7 +228,14 @@ void Thunderloop::runLoop()
...
             *(robot_status_.mutable_thunderloop_status()) = thunderloop_status_;
             *(robot_status_.mutable_motor_status())       = motor_status_;
+            *(robot_status_.mutable_power_status())       = power_status_;
             *(robot_status_.mutable_jetson_status())      = jetson_status_;
+
+            // Update Redis
+            redis_client_->set("/battery_voltage",
+                               std::to_string(power_status_.battery_voltage()));
+            redis_client_->set("/current_draw",
+                               std::to_string(power_status_.current_draw()));
```
Comment (nimazareian) [inline src/software/jetson_nano/thunderloop.cpp]: Can we use the constants here as well?

File: src/software/jetson_nano/thunderloop.h [hunk 12]
```diff
@@ -97,13 +96,14 @@ class Thunderloop
...
     RobotConstants_t robot_constants_;
+    Angle current_orientation_;
     int robot_id_;
     int channel_id_;
     std::string network_interface_;
     int loop_hz_;
 
     // 50 millisecond timeout on receiving primitives before we emergency stop the robots
-    const double PRIMITIVE_MANAGER_TIMEOUT_NS = 50.0 * MILLISECONDS_PER_NANOSECOND;
+    const double PRIMITIVE_MANAGER_TIMEOUT_NS = 500.0 * NANOSECONDS_PER_MILLISECOND;
```
Comment (nimazareian) [inline src/software/jetson_nano/thunderloop.h]: The comment should be updated to 500 millisecond

File: src/software/jetson_nano/thunderloop.cpp [hunk 13]
```diff
@@ -157,34 +160,65 @@ void Thunderloop::runLoop()
...
 
-                // If we haven't received a primitive in a while, override the
-                // current_primitive_ with an estop primitive.
                 if (nanoseconds_elapsed_since_last_primitive >
                     static_cast<long>(PRIMITIVE_MANAGER_TIMEOUT_NS))
                 {
-                    primitive_.Clear();
-                    *(primitive_.mutable_estop()) = emergency_stop_override;
+                    primitive_executor_.clearCurrentPrimitive();
                 }
```
Comment (nimazareian) [inline src/software/jetson_nano/thunderloop.cpp:177]: Could you add a todo for #2726 either in this file or in primitive executor? 
We should also not move if a World hasn't been received in a while.

**Review by nimazareian — COMMENTED** [review]
Overall looks good to me. Just had a couple of comments.
Excited to get this merged soon.

File: src/software/jetson_nano/services/network.cpp (hunk 9 above)
Comment (LiCody) [reply src/software/jetson_nano/services/network.cpp]: I believe that implementation will over count the number of primitives lost when multiple primitives are lost sequentially. Assuming the callback is always eventually called the case you mentioned would be handled correctly with the current implementation

File: src/shared/2021_robot_constants.cpp (hunk 1 above)
Comment (LiCody) [reply src/shared/2021_robot_constants.cpp:18]: Unsure if these still need to updated so leaving open

File: src/software/jetson_nano/services/network.cpp (hunk 9 above)
Comment (nimazareian) [reply src/software/jetson_nano/services/network.cpp]: Can we assume that the callback is always called though? If one primitive can be lost, is it not possible for multiple primitives to be lost back to back?

File: src/shared/2021_robot_constants.cpp (hunk 5 above)
Comment (LiCody) [reply src/shared/2021_robot_constants.cpp:28]: agreed I believe this is being negated in implementation but the constant makes more sense to be positive. This is currently not being used will investigate and follow up

File: src/software/jetson_nano/services/network.cpp (hunk 9 above)
Comment (LiCody) [reply src/software/jetson_nano/services/network.cpp]: re reading the code your implementation wont over count and is fine for the assumption of callback being dropped

File: src/software/jetson_nano/services/network.cpp (hunk 4 above)
Comment (LiCody) [reply src/software/jetson_nano/services/network.cpp:51]: Good idea tagging an issue

File: src/software/jetson_nano/display/screens/home_screen.py (hunk 7 above)
Comment (LiCody) [reply src/software/jetson_nano/display/screens/home_screen.py:25]: seems like a decent difficulty 1 ticket. tagged a ticket

File: src/shared/2021_robot_constants.cpp (hunk 5 above)
Comment (LiCody) [reply src/shared/2021_robot_constants.cpp:28]: So indefinite dribbler speed is unused and should be negated and max force dribbler speed is only used by a few fsms. going to do a quick check in another pr over gameplay changes from robocup. For this I'm just going to add a comment about how negative is sucking the ball in

File: src/software/thunderscope/robot_diagnostics/drive_and_dribbler_widget.py [hunk 14]
```diff
@@ -33,18 +33,11 @@ def __init__(self, proto_unix_io):
 
         # Initialize tab screen
         self.tabs = QTabWidget()
-        per_wheel_tab = QWidget()
         direct_velocity_tab = QWidget()
 
         # Add tabs
-        self.tabs.addTab(per_wheel_tab, "Direct Per-Wheel Control")
         self.tabs.addTab(direct_velocity_tab, "Direct Velocity Control")
```
Comment (nimazareian) [inline src/software/thunderscope/robot_diagnostics/drive_and_dribbler_widget.py:39]: Curious why you removed the per wheel tab? I've never used it (didn't know we had it), but I guess it could be useful for Mech/Elec.

Comment (akhilveeraghanta) [reply src/software/thunderscope/robot_diagnostics/drive_and_dribbler_widget.py:39]: IIRC it was removed at robocup because we don't have per-wheel ramping and that tab messes the motors up if you slide the rpm too fast. So if you need per-wheel rpm control, you'd need to support per-wheel ramping before enabling this tab

Comment (nimazareian) [reply src/software/thunderscope/robot_diagnostics/drive_and_dribbler_widget.py:39]: That makes more sense.
@LiCody I think it might be worthwhile to add a task for supporting per wheel velocity in robot diagnostic. You could probably link to a branch which has this implementation in it, and note that ramping is required before it can be tested.

Comment (LiCody) [reply src/software/thunderscope/robot_diagnostics/drive_and_dribbler_widget.py:39]: created a ticket

**Review by nimazareian — APPROVED** [review]
LGTM! 🎮

**Review by itsarune — APPROVED** [review]
